Add Section 2.3 (Lean statements) (#607)
fpvandoorn/carleson: paper/main.tex

which may or may not converge for a given $x\in\real$.
\end{definition}

The metric space Carleson theorem in \cite{CarlesonBlueprint} is a restricted weak type estimate for a generalized Carleson operator. One of the generalizations here is that the domain of the functions operated on can be any \emph{doubling metric measure space} $(X,\rho,\mu, a)$.

> \begin{definition}\label{def:doubling}

Reviewer: Make this longer, to disambiguate from a doubling measure? (optional)

The pieces are categorized in one of three types, for each of which a separate analysis is required to establish a bound.
These three branches of the proof are largely independent and handled in separate sections 5, 6, and 7 of the blueprint \cite{CarlesonBlueprint}.

\subsection{Stating the main results in Lean}\label{subsec:lean-statements}

> Starting from the material already available in Lean's mathematical library \mathlib, it is easy to state the classical Carleson theorem in Lean\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Classical/ClassicalCarleson.lean#L132}{\extlink}.

Reviewer: ```suggestion

Starting from the material already available in Lean's mathematical library \mathlib, it is easy to state the classical Carleson theorem in Lean\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Classical/ClassicalCarleson.lean#L132}{\extlink}.
```

Reviewer: Also, we should unify how we do external links. But that can happen in a follow-up.

Author: I'll add this to the review checklist.

\begin{lstlisting}[mathescape]
theorem classical_carleson {f : ℝ → ℂ} (cont_f : Continuous f)

>   (periodic_f : f.Periodic (2 * π)) :

Reviewer: minor tweak
```suggestion
theorem classical_carleson {f : ℝ → ℂ}
  (cont_f : Continuous f) (periodic_f : f.Periodic (2 * π)) :
```

  ∀$^m$ x, Filter.Tendsto (partialFourierSum · f x) Filter.atTop (nhds (f x))
\end{lstlisting}

> The statement of the hypotheses ($f : \R \to \C$ is a continuous $2\pi$-periodic function) closely resembles natural language, but understanding the conclusion requires some familiarity with Lean and Mathlib. There, $\forall^m x$ means `for every real $x$ away from a set of measure zero', and

Reviewer: ```suggestion
The statement of the hypotheses ($f\colon\real\to\complex$ is a continuous $2\pi$-periodic function) closely resembles natural language, but understanding the conclusion requires some familiarity with Lean and \mathlib. There, $\forall^m x$ means `for every real number $x$ away from a set of measure zero', and
```

> \lean{Filter.Tendsto (partialFourierSum · f x) Filter.atTop (nhds (f x))} is Lean's idiomatic way to express that the sequence $S_N f(x)$ of $N$-th partial Fourier sums converges to $f(x)$ when $N$ tends to infinity (see \cite[\S11.1]{mathematics_in_lean} for more information on Mathlib's use of filters to represent limits).

Reviewer: ```suggestion
\lean{Filter.Tendsto (partialFourierSum · f x) Filter.atTop (nhds (f x))} is Lean's idiomatic way to express that the sequence $S_N f(x)$ of $N$-th partial Fourier sums converges to $f(x)$ when $N$ tends to infinity (see \cite[\S11.1]{mathematics_in_lean} for more information on \mathlib's use of filters to represent limits).
```

> The definition \lean{partialFourierSum}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L32}{\extlink} encodes the $N$-th partial Fourier sum of $f : \R \to \C$. Here, \lean{Finset.Icc}\footnote{In any preorder $\alpha$, one can define intervals which on each side can be either open (o), closed (c) or infinite (i). For instance, \lean{Set.Ico a b} denotes the inverval $[a, b)$ and \lean{Set.Iic b} denotes $(-\infty, b]$. If moreover $\alpha$ is a \lean{LocallyFiniteOrder} (i.e., all bounded intervals in $\alpha$ are finite), \lean{Finset.Icc, Finset.Ico, Finset.Ioc} and \lean{Finset.Ioo} provide \lean{Finset} versions of these intervals.

Reviewer: style nit:
```suggestion
The definition \lean{partialFourierSum}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L32}{\extlink} encodes the $N$-th partial Fourier sum of $f\colon\real\to\complex$. Here, \lean{Finset.Icc}\footnote{In any preorder $\alpha$, one can define intervals which on each side can be either open (o), closed (c) or infinite (i). For instance, \lean{Set.Ico a b} denotes the inverval $[a, b)$ and \lean{Set.Iic b} denotes $(-\infty, b]$. If moreover $\alpha$ is a \lean{LocallyFiniteOrder} (i.e., all bounded intervals in $\alpha$ are finite), \lean{Finset.Icc, Finset.Ico, Finset.Ioc} and \lean{Finset.Ioo} provide \lean{Finset} versions of these intervals.
```

Reviewer: Additional comments: should this briefly explain what a `Finset` is? (You're doing so in §5, perhaps this can move up then.)
(I'm also not a huge fan of "one can define" to describe something done, but that's getting really persnickety.)

Author: I put this intentionally as a footnote for someone with a bit more familiarity with Lean. I don't want to deviate too much from describing the statements.

} denotes the finite set of integers $\{-N, -N + 1, \cdots, N\}$, \lean{fourierCoeffOn Real.two_pi_pos f n} is the

> \lean{n}-th Fourier coefficient of $f$ on $[0, 2\pi]$, and \lean{fourier n (x : AddCircle (2 * π))} denotes $e^{inx}$.

Reviewer: ```suggestion
$n$-th Fourier coefficient of $f$ on $[0, 2\pi]$, and \lean{fourier n (x : AddCircle (2 * π))} denotes $e^{inx}$.
```

\begin{lstlisting}
def partialFourierSum (N : ℕ) (f : ℝ → ℂ) (x : ℝ) : ℂ :=
  ∑ n ∈ Finset.Icc (-(N : ℤ)) N,
    fourierCoeffOn Real.two_pi_pos f n * fourier n (x : AddCircle (2 * π))
\end{lstlisting}

> On the other hand, in order to formalize the \emph{statements} of the metric and linearized versions of Carleson's theorem, we first needed to formalize a good amount of new concepts, starting with the definition of doubling metric measure space (see Definition \ref{def:doubling}), which we denote by \lean{MeasureTheory.DoublingMeasure}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L87}{\extlink}. The constant $A$ appearing in this definition will typically be $2^a$.

Reviewer: ```suggestion
On the other hand, in order to formalize the \emph{statements} of the metric and linearized versions of Carleson's theorem, we first needed to formalize a good amount of new concepts, starting with the definition of a doubling metric measure space (see Definition \ref{def:doubling}), which we denote by \lean{MeasureTheory.DoublingMeasure}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L87}{\extlink}. The constant $A$ appearing in this definition will typically be $2^a$.
```

Author: Are you sure? This does not sound good to me.

  CompleteSpace X, LocallyCompactSpace X, MeasureSpace X, BorelSpace X,
  IsLocallyFiniteMeasure (volume : Measure X),
  Measure.IsDoubling (volume : Measure X) A, NeZero (volume : Measure X) where
\end{lstlisting}

Next, we need to introduce the family $\Mf$ of compatible functions satisfying the properties \eqref{osccontrol}, \eqref{firstdb}, \eqref{monotonedb}, \eqref{seconddb}, and \eqref{thirddb}. We do this in two steps, by first creating a class
\lean{FunctionDistances}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L102}{\extlink} which bundles a family of continuous functions and then endowing these functions with the five properties in the class \lean{CompatibleFunctions}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L157}{\extlink}. The assumption that the family is cancellative is added as a \lean{Prop}-valued class \lean{IsCancellative}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L181}{\extlink}.

> Something to note here is that we created type synonyms to be able to endow $\Theta X$ with different metric space structures, each coming from a ball in $X$, and we provided the custom notation \lean{dist_\{x, r\}} for the distance corresponding to the ball $B(x, r)$.

Reviewer: In the blueprint, we just write $\Theta$. Otherwise this should be `Theta X` in Lean code. What do you think?

Author: I switched to $\Theta$.

def WithFunctionDistance (x : X) (r : ℝ) := Θ X

instance {x : X} {r : ℝ} [d : FunctionDistances k X] :
  PseudoMetricSpace (WithFunctionDistance x r) := d.metric x r

notation3 "dist_{" x ", " r "}" => @dist (WithFunctionDistance x r) _
\end{lstlisting}

With this notation, the Lean statement of the compatibility properties is quite close to the informal one. For instance, equation \eqref{monotonedb} is written as
\begin{lstlisting}
cdist_mono {x₁ x₂ : X} {r₁ r₂ : ℝ} {f g : Θ}
  (h : ball x₁ r₁ ⊆ ball x₂ r₂) : dist_{x₁, r₁} f g ≤ dist_{x₂, r₂} f g
\end{lstlisting}

The class \lean{IsOneSidedKernel}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L240}{\extlink} captures the notion of one-sided Calder\'on--Zygmund kernel (Definition \ref{def:one-sided-kernel}).

\begin{lstlisting}
class IsOneSidedKernel (a : outParam ℕ) (K : X → X → ℂ) : Prop where
  measurable_K : Measurable (uncurry K)
  norm_K_le_vol_inv (x y : X) : ‖K x y‖ ≤ C_K a / Real.vol x y
  norm_K_sub_le {x y y' : X} (h : 2 * dist y y' ≤ dist x y) : ‖K x y - K x y'‖ ≤
    (dist y y' / dist x y) ^ (a : ℝ)⁻¹ * (C_K a / Real.vol x y)
\end{lstlisting}

> To avoid constant repetition of hypotheses, we create a class \lean{KernelProofData}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L251}{\extlink} which contains data used in most of chapters 2 through 7 of the blueprint: a doubling measure on $X$, a natural number greater than three, a collection of compatible functions, and a one-sided Calder\'on--Zygmund kernel.

Reviewer: Can you insert a forward reference to 5.2 (the ProofData pattern) here?

\lean{nontangentialOperator}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L201}{\extlink}.

\begin{lstlisting}
theorem metric_carleson {X : Type*} {a : ℕ} [MetricSpace X] {q q' : ℝ≥0}
  {F G : Set X} {K : X → X → ℂ} [KernelProofData a K] {f : X → ℂ}
  [IsCancellative X (defaultτ a)] (hq : q ∈ Ioc 1 2)
  (hqq' : q.HolderConjugate q') (mF : MeasurableSet F) (mG : MeasurableSet G)
  (mf : Measurable f) (nf : (‖f ·‖) ≤ F.indicator 1)
  (hT : HasBoundedStrongType (nontangentialOperator K · ·) 2 2 volume volume (C_Ts a)) :
  ∫⁻ x in G, carlesonOperator K f x ≤
    C1_0_2 a q * volume G ^ (q' : ℝ)⁻¹ * volume F ^ (q : ℝ)⁻¹
\end{lstlisting}

%\begin{lstlisting}[mathescape]
%	def nontangentialOperator (K : X → X → ℂ) (f : X → ℂ) (x : X) : ℝ≥0∞ :=
%	$\sqcup$ (R₂ : ℝ) (R₁ ∈ Ioo 0 R₂) (x' ∈ ball x R₁),
%	‖∫ y in Annulus.oo x' R₁ R₂, K x' y * f y‖$_e$
%\end{lstlisting}

> The \lean{carlesonOperator}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L218}{\extlink} is defined as a special case of the

Reviewer: General comment: should we capitalize these references, as in Equation~(19)?

Author: I think not if it is plural as in this case, but I don't have a strong preference.

> \lean{linearizedCarlesonOperator}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L212}{\extlink}; these definitions correspond to those introduced in equations (\ref{def-main-op}) and (\ref{def-T-r}), respectively.

Reviewer: LaTeX nit: \eqref can help us here: `eqref{foo}` is equivalent to `(ref{foo})` but nicer.
(I personally also like prefixing my equation labels with `eq:`, so I don't accidentally reference a definition as an equation or similarly. Personal style.)

In \lean{carlesonOperatorIntegrand}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L206}{\extlink} below, \lean{Set.Annulus.oo x R₁ R₂}\href{https://github.com/fpvandoorn/carleson/blob/f2f614db47f1a30e3fceaa9c2e1f99ff356dd577/Carleson/Defs.lean#L44}{\extlink} denotes the set of real numbers $y$ such that $R_1 < \text{dist}(x, y) < R_2$.

\begin{lstlisting}[mathescape]
def carlesonOperatorIntegrand [FunctionDistances ℝ X] (K : X → X → ℂ)
    (θ : Θ X) (R₁ R₂ : ℝ) (f : X → ℂ) (x : X) : ℂ :=
  ∫ y in Annulus.oo x R₁ R₂, K x y * f y * exp (I * θ y)

def linearizedCarlesonOperator [FunctionDistances ℝ X] (Q : X → Θ X)
    (K : X → X → ℂ) (f : X → ℂ) (x : X) : ℝ≥0∞ :=
  $\sqcup$ (R₁ : ℝ) (R₂ : ℝ) (_ : 0 < R₁) (_ : R₁ < R₂),

>     ‖carlesonOperatorIntegrand K (Q x) R₁ R₂ f x‖$_e$

Reviewer: This notation (enorms) is not explained yet. How about leaving a forward reference, and a short explanation?

def carlesonOperator [FunctionDistances ℝ X] (K : X → X → ℂ) (f : X → ℂ)
    (x : X) : ℝ≥0∞ :=

>   $\sqcup$ (θ : Θ X), linearizedCarlesonOperator (fun _ ↦ θ) K f x

Reviewer: disjoint union, right? I'm not sure if the math section explains this --- can you add a TODO (or check)?

Author: I added a TODO in the math section.
